lib/pages/search/server_search_controller.dart

```diff
@@ -14,34 +19,114 @@ class ServerSearchController with SearchDebouncerMixin {
     Right(ServerSearchInitial()),
   );
 
-  String keyword = '';
+  static const int _limitServerSideSearchFilter = 20;
+
+  final isLoadingMoreNotifier = ValueNotifier<bool>(false);
+
+  final scrollController = ScrollController();
+
+  StreamSubscription? _searchSubscription;
+
+  String? _nextBatch;
+
+  ServerSideSearchCategories? _searchCategories;
 
   void init() {
     initializeDebouncer((searchTerm) {
+      updateSearchCategories(searchTerm);
       if (searchTerm.isNotEmpty) {
-        keyword = searchTerm;
-        searchUnencryptedMessages(searchTerm);
+        searchUnencryptedMessages();
       } else {
         serverSearchNotifier.value = Right(ServerSearchInitial());
       }
     });
+    scrollController.addLoadMoreListener(loadMore);
+    serverSearchNotifier.addListener(_listenSearchResult);
   }
 
   void dispose() {
     super.disposeDebouncer();
+    scrollController.dispose();
+    serverSearchNotifier.dispose();
+    _searchSubscription?.cancel();
+    isLoadingMoreNotifier.dispose();
+    resetNextBatch();
+    _searchCategories = null;
   }
 
-  void searchUnencryptedMessages(String searchTerm) {
-    _serverSearchInteractor
-        .execute(
+  void _listenSearchResult() {
+    serverSearchNotifier.value.fold(
+      (failure) => resetNextBatch(),
+      (success) {
+        if (success is ServerSearchChatSuccess) {
+          updateNextBatch(success.nextBatch);
+        }
+      },
+    );
+  }
+
+  void resetNextBatch() {
+    _nextBatch = null;
+  }
+
+  void updateNextBatch(String? newBatch) {
+    _nextBatch = newBatch;
+  }
+
+  void updateSearchCategories(String searchTerm) {
+    _searchCategories = ServerSideSearchCategories(
       searchTerm: searchTerm,
+      searchFilter: SearchFilter(
+        limit: _limitServerSideSearchFilter,
+      ),
+    );
+  }
+
+  void searchUnencryptedMessages() {
+    _searchSubscription = _serverSearchInteractor
+        .execute(
+      searchCategories: _searchCategories!,
     )
-        .listen((searchResult) {
-      serverSearchNotifier.value = searchResult;
-    });
+        .listen(
+      (searchResult) {
+        serverSearchNotifier.value = searchResult;
+      },
+      onDone: () {
+        scrollController.tryLoadMoreIfNeeded(loadMore);
```

Comment:
why need try to loadMore in `onDone`?

Reply:
IDK why Minh created it. Just I tested it when removed it and works well. IMO remove it

Reply:
- Currently, we are adding a `limit` to the default server-side search of 20. So the returned list of results is quite enough to display the UI so there is no need for that.
- For me, that function is not necessary, the user can continue scrolling to load more, so it should not be automatically triggered

Reply:
I am concerned about the loop call.